File: src/main/java/org/owntracks/android/support/receiver/Parser.java
package org.owntracks.android.support.receiver;

import android.content.Context;
import android.support.annotation.NonNull;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.owntracks.android.messages.MessageBase;
import org.owntracks.android.messages.MessageEncrypted;
import org.owntracks.android.support.EncryptionProvider;

import java.io.IOException;
import java.io.InputStream;

public class Parser {
    private static final String TAG = "Parser";
    static ObjectMapper mapper;

    public static void initialize(Context c) {
        mapper = new ObjectMapper();

    }



    private static MessageBase pipelineDecrypt(MessageBase m) throws IOException, EncryptionException {
        if(m instanceof MessageEncrypted) {
            if(!EncryptionProvider.isPayloadEncryptionEnabled())
                throw new EncryptionException("received encrypted message but payload encryption is not enabled");
            return mapper.readValue(EncryptionProvider.decrypt(((MessageEncrypted) m).getData()), MessageBase.class);
        }
        return m;
    }

    private static MessageBase pipelineDeserialize(InputStream input) throws IOException, EncryptionException {
        return mapper.readValue(input, MessageBase.class);
    }
    private static MessageBase pipelineDeserialize(@NonNull byte[] input) throws IOException, EncryptionException {
        return mapper.readValue(input, MessageBase.class);
    }

    public static MessageBase deserializeSync(@NonNull InputStream input ) throws IOException, EncryptionException {
        return pipelineDecrypt(pipelineDeserialize(input));
    }

    public static MessageBase deserializeSync(@NonNull byte[] input) throws IOException, EncryptionException {
        return pipelineDecrypt(pipelineDeserialize(input));
    }

    private static String pipelineSerialize(@NonNull MessageBase input) throws IOException, EncryptionException {
        return mapper.writeValueAsString(input);
    }

    private static String pipelineEncrypt(@NonNull String input) throws IOException, EncryptionException {
        if(EncryptionProvider.isPayloadEncryptionEnabled()) {
            MessageEncrypted m = new MessageEncrypted();
            m.setdata(EncryptionProvider.encrypt(input));
            return mapper.writeValueAsString(input);
        }
        return input;
    }

    public static String serializeSync(@NonNull MessageBase message) throws IOException, EncryptionException {
        return pipelineEncrypt(pipelineSerialize(message));
    }

    public static class EncryptionException extends Exception {
        public EncryptionException(String s) {
            super(s);
        }
    }
}

File: project/app/src/main/java/org/owntracks/android/model/Parser.kt
package org.owntracks.android.model

import java.io.IOException
import java.io.InputStream
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.owntracks.android.model.messages.MessageBase
import org.owntracks.android.model.messages.MessageEncrypted
import org.owntracks.android.model.messages.MessageUnknown

@Singleton
class Parser @Inject constructor(private val encryptionProvider: EncryptionProvider?) {
  private val defaultMapper = Json {
    prettyPrint = true
    ignoreUnknownKeys = true
    classDiscriminator = "_type"
  }

  private val arrayCompatMapper = Json {
    ignoreUnknownKeys = true
    isLenient = true
    classDiscriminator = "_type"
  }

  @Throws(IOException::class)
  fun toUnencryptedJsonPretty(message: MessageBase): String {
    return defaultMapper.encodeToString(message).replace("\\r\\n".toRegex(), "\n")
  }

  @Throws(IOException::class)
  fun toJsonPlain(message: MessageBase): String {
    return defaultMapper.encodeToString(message)
  }

  @Throws(IOException::class)
  fun toUnencryptedJsonBytes(message: MessageBase): ByteArray {
    return defaultMapper.encodeToString(message).toByteArray()
  }

  @Throws(IOException::class)
  fun toJson(message: MessageBase): String {
    return encryptString(toJsonPlain(message))
  }

  @Throws(IOException::class)
  fun toJsonBytes(message: MessageBase): ByteArray {
    return encryptBytes(toUnencryptedJsonBytes(message))
  }

  @Throws(IOException::class, EncryptionException::class)
  fun fromJson(input: String): MessageBase {
    return decrypt(defaultMapper.decodeFromString<MessageBase>(input))
  }

  @Throws(IOException::class)
  fun fromUnencryptedJson(input: ByteArray): MessageBase =
      defaultMapper.decodeFromString(input.toString(Charsets.UTF_8))

  // Accepts {plain} as byte array
  @Throws(IOException::class, EncryptionException::class)
  fun fromJson(input: ByteArray): MessageBase =
      try {
        decrypt(fromUnencryptedJson(input))
      } catch (e: Exception) {
        MessageUnknown
      }

  // Accepts 1) [{plain},{plain},...], 2) {plain}, 3) {encrypted, data:[{plain}, {plain}, ...]} as
  // input stream
  @Throws(IOException::class, EncryptionException::class)
  fun fromJson(input: InputStream): Array<MessageBase> {
    return decrypt(arrayCompatMapper.decodeFromString<Array<MessageBase>>(input.bufferedReader().use { it.readText() }))
  }

  @Throws(IOException::class, EncryptionException::class)
  private fun decrypt(a: Array<MessageBase>?): Array<MessageBase> {
    // Recorder compatibility, encrypted messages with data array
    if (a == null) throw IOException("null array")
    return if (a.size == 1 && a[0] is MessageEncrypted) {
      if (encryptionProvider == null || !encryptionProvider.isPayloadEncryptionEnabled) {
        throw EncryptionException(
            "received encrypted message but payload encryption is not enabled", null)
      }
      val encryptedMessage = a[0] as MessageEncrypted
      val decrypted: String = encryptionProvider.decrypt(encryptedMessage.data)
      fromJson(decrypted.byteInputStream())
    } else {
      a
    }
  }

  @Throws(EncryptionException::class)
  private fun decrypt(message: MessageBase): MessageBase {
    return if (message is MessageEncrypted) {
      if (encryptionProvider == null || !encryptionProvider.isPayloadEncryptionEnabled) {
        throw EncryptionException(
            "received encrypted message but payload encryption is not enabled", null)
      }
      val decrypted: String = encryptionProvider.decrypt(message.data)
      fromJson(decrypted)
    } else {
      message
    }
  }

  @Throws(IOException::class)
  private fun encryptString(string: String): String {
    return if (encryptionProvider != null && encryptionProvider.isPayloadEncryptionEnabled) {
      val encrypted = MessageEncrypted()
      encrypted.data = encryptionProvider.encrypt(string)
      toJsonPlain(encrypted)
    } else {
      string
    }
  }

  @Throws(IOException::class)
  private fun encryptBytes(bytes: ByteArray): ByteArray {
    return if (encryptionProvider != null && encryptionProvider.isPayloadEncryptionEnabled) {
      val encrypted = MessageEncrypted()
      encrypted.data = encryptionProvider.encrypt(bytes)
      toUnencryptedJsonBytes(encrypted)
    } else {
      bytes
    }
  }

  class EncryptionException internal constructor(s: String, cause: Throwable?) :
      Exception(s, cause) {
    constructor(s: String) : this(s, null)
  }
}
